fiftyone/core/collections.py

```diff
@@ -9044,6 +9045,35 @@ def values(
         Returns:
             the list of values
         """
+
+        # If we do not need to follow insertion order, we can potentially use a
+        # covered index query to get the values directly from the index and
+        # avoid a COLLSCAN.
+        if not _enforce_natural_order:
+            field = None
+            if isinstance(field_or_expr, str):
+                field = field_or_expr
+            elif isinstance(field_or_expr, list) and len(field_or_expr) == 1:
+                field = field_or_expr[0]
+
+            # Checking _is_full_collection last as it could potentially be expensive.
+            # In the future, we might be less restrictive, but that needs
+            # further investigation.
+            if (
+                field in ["id", "_id", "filepath"]
+                and self._is_full_collection()
+            ):
+                try:
+                    return foo.get_indexed_values(
+                        self._dataset._sample_collection,
+                        field,
+                        values_only=True,
+                    )
+                except Exception as e:
```

Comment:
Can we not use the `try`-`except` pattern here please? 

Here's my thinking: core methods need to be as fast as possible, and they need to work when used correctly or not work when used incorrectly. Catching an exception and gracefully continuing implies that we first tried something (potentially expensive), but it failed, so we're falling back to some other potentially expensive thing. It's in the best interest of the caller to just see the exception so that they know they shouldn't be using the method differently.

In other words, we shouldn't be shipping some feature that might log warnings here, because by definition that thing is doing more work than it could be.

Reply:
Doesn't matter too much because it's a private usage, but yeah that should be fine I'd think.

Reply:
>  core methods need to be as fast as possible, and they need to work when used correctly or not work when used incorrectly. 
the `try: except:` allows us to attempt to run it without checking the indexes manually first. Since checking indexes requires elevated mongo permissions, I didn't want to potentially trigger that error in a values() call. 
I know that the sidebar frequently checks the indexes, but an exclusive sdk user potentially may have never hit those codpaths and needed those permissions .

If it's not in the index, the error will throw immediately (the find will not attempt to execute)

Reply:
I realize that in this particular case, the idea is "check to see if there is a covered index, and if there is, then use a more optimal code path to get the values". So in fact the pathway through the function is not deterministic, it may go one way or another. 

My point is that I'd like to see the code implemented more like this:

```py
if not _enforce_natural_order and _has_covered_indexes(...):
    return _get_indexed_values(...)
```

Then it is clear what is going on.

Reply:
In my opinion, the semantics of passing `_enforce_natural_order=False` are "get me the values in any order you want". So if we can get the values in natural order, why are we logging a warning?

If the goal is to provide a hint to the user that the method could be optimized by adding a certain index, then that seems more like a `debug()` use case to me.

Reply:
I am fine changing it back to debug.

Is `_has_covered_indexes(...)` a blocking requirement?

Reply:
Also, suppose I accidentally contribute a bug to `get_indexed_values()` where it no longer worked. The way it is invoked by `values()` now with the naked `try`-`except` would suppress the real bug and turn it into a warning that isn't the user's fault!

Reply:
> the try: except: allows us to attempt to run it without checking the indexes manually first. Since checking indexes requires elevated mongo permissions, I didn't want to potentially trigger that error in a values() call.
I know that the sidebar frequently checks the indexes, but an exclusive sdk user potentially may have never hit those codpaths and needed those permissions .

I appreciate the careful thinking here, but given the following:
- creating datasets requires the ability to create and index collections
- as you mention, the App's sidebar does lots of index checks

the reality is that all user personas need to be able to CRUD indexes